### ISSUE TITLE
rust: std::os::unix::fs::FileTypeExt does not work

### ISSUE THREAD

**Issue by Crestwave** [haikuports/haikuports#3264]
#3251

**Comment by Crestwave** [comment]
The problem seems to be that [this line](https://github.com/BurntSushi/ripgrep/blob/743d64f2e4093a3302895e128fbbc58e6fb8ed18/src/search.rs#L312) (`if subject.is_stdin() {`) always fails, so it doesn't even *try* to read from `stdin`. The function it calls is from [subject.rs](https://github.com/BurntSushi/ripgrep/blob/743d64f2e4093a3302895e128fbbc58e6fb8ed18/src/subject.rs#L126):

```rust
    pub fn is_stdin(&self) -> bool {
        self.dent.is_stdin()
    }
```

**Comment by Crestwave** [comment]
It seems that the problem is simply in detecting `stdin`; this can be worked around by passing `-` to explicitly specify it.

**Comment by Crestwave** [comment]
My only guess as to the cause is that the value of https://github.com/rust-lang/libc/blob/ff63cb517872d95bdba9960d4d200d024c1805a0/src/unix/haiku/mod.rs#L573-L575 and others seem wrong, but I don't have experience in these things and am probably not making any sense. Any idea @nielx?

**Comment by nielx** [comment]
The system defines are at [sys/stat.h](https://github.com/haiku/haiku/blob/7c32619c7fa898b2a081f1bc7c977fef431576d9/headers/posix/sys/stat.h#L56-L63). They are formatted as raw octal values, and if you convert them, some do not match with the ones in libc. 

The values for S_IFIFO and S_IFCHR seem to be wrong. The other ones in the S_IF* group are correct.

**Comment by Crestwave** [comment]
Yay, I guessed correctly; those are the ones I tested and suspected were wrong. `ripgrep` uses `S_IFIFO`; fixing that value should fix the stdin detection. So it's 4096 and 8192 for `S_IFIFO` and `S_IFCHR` respectively, right?

**Comment by nielx** [comment]
Yes, that's correct. 

I have prepared a patch for libc in this tree: https://github.com/nielx/libc/tree/fix/stat-constants
I don't know what the turnaround time is on increasing the libc version though, so for this recipe you might want to patch the constant directly.

**Comment by nielx** [comment]
For reference: https://github.com/rust-lang/libc/pull/1537

**Comment by Crestwave** [comment]
It's part of the stdlib, though. How would I patch it directly?

**Comment by nielx** [comment]
Oh sorry, I understood ripgrep uses S_IFIFO directly

**Comment by Crestwave** [comment]
Ah, I guess I didn't word it well. It uses std::os::unix::fs::FileTypeExt's `is_fifio()` with uses `S_IFIFO`.

**Comment by nielx** [comment]
Right, in this case the issue will be fixed when ripgrep is built with a newer version of rust that includes the patch. I think that will be rust 1.39.0 or 1.40.0. Since there is a workaround currently, I would not go and hack ripgrep to workaround the faulty constant.

**Comment by Crestwave** [comment]
Resolved with #4491
